### PR TITLE
feat: use custom help classes

### PR DESCRIPTION
**Pull request by chadian**
This pull request adds the functionality for creating custom help classes and to find if a custom help class has been specified in a CLI's package.json. The change to the package.json is included from the latest change to `oclif/config` which adds `helpClass` to the oclif section of a CLI's package.json:
* [Bump in `@oclif/config`](https://github.com/oclif/plugin-help/pull/62/files#diff-8ee2343978836a779dc9f8d6b794c3b2R98
) in yarn.lock
* via https://github.com/oclif/config/pull/103

Custom help classes can be an extension of the default the `BaseClass` exported from this package which specifies the mandatory methods to be implemented for a help class to work with the oclif when it needs to show help.  The `Help` class exported from this package can also be extended if only specific methods want to be overridden from it.

The `getHelpClass` will be used by `@oclif/command` to find a corresponding help class and use that, if it exists, otherwise falling back to using the default help class exported from `@oclif/plugin-help`.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -52,12 +61,9 @@ export default class Help {
     const subject = getHelpSubject(argv)
 
     let command: Config.Command | undefined
+    let topic: Config.Topic | undefined
     if (subject) {
       command = this.config.findCommand(subject)
-    }
-
-    let topic: Config.Topic | undefined
-    if (subject && !command) {
```
Comment (amphro) [inline src/index.ts:60]: Is there any point to find the topic if the command doesn't exist?

File: src/index.ts [hunk 2]
```diff
@@ -16,16 +16,12 @@ const {
 } = chalk
 
 export interface HelpOptions {
-  all?: boolean;
   maxWidth: number;
   stripAnsi?: boolean;
+  all?: boolean;
```
Comment (amphro) [inline src/index.ts:21]: Why move this down?

File: src/index.ts (hunk 1 above)
Comment (chadian) [reply src/index.ts:60]: Agreed, that's the  way it was before. The diff makes it tricky to see the change, but I moved both `topic` and `command` together when there is a `subject`:
https://github.com/oclif/plugin-help/blob/7211fd91b2dd020a58f3659cce35bfee04af0c76/src/index.ts#L63-L68

Does that fix what what you're describing?

File: src/index.ts (hunk 2 above)
Comment (chadian) [reply src/index.ts:21]: 👍 Shuffled in the refactor, I'll put it back.

File: src/index.ts (hunk 1 above)
Comment (amphro) [reply src/index.ts:60]: It just seemed before it was checking the `!command` and now you are not. So there could be a case where the command is undefined but you still find the topic?

Comment (chadian) [reply src/index.ts:60]: I've changed this a bit now. I've added a test to track the old routing for the previous methods, and made the dependencies for what is shown happen in lock step. I think this makes it easier to see.

Yes, there is a case where the command is undefined and we still have a topic to show. This case is the last chance to find a match for the `subject` before an error is thrown.

File: src/index.ts [hunk 3]
```diff
@@ -34,77 +31,157 @@ function getHelpSubject(args: string[]): string | undefined {
...
     const title = command.description && this.render(command.description).split('\n')[0]
     if (title) console.log(title + '\n')
-    console.log(this.command(command))
+    console.log(this.formatCommand(command))
+    console.log('')
+
+    if (topics.length > 0) {
+      console.log(this.formatTopics(topics))
+      console.log('')
+    }
```
Comment (RasPhilCo) [inline src/index.ts:110]:
```suggestion
    if (commands.length > 0) {
      console.log(this.formatCommands(commands))
      console.log('')
    }
```

File: src/index.ts [hunk 4]
```diff
@@ -58,8 +58,32 @@ export abstract class HelpBase {
 export default class Help extends HelpBase {
   render: (input: string) => string
 
+  /*
+   * _topics is to work around Config.topics mistakenly including commands that do
+   * not have children, as well as topics. A topic has children, either commands or other topics. When
+   * this is fixed upstream config.topics should return *only* topics with children,
+   * and this can be removed.
+   */
+  protected get _topics(): Config.Topic[] {
```
Comment (chadian) [inline src/index.ts:67]: make private

File: src/index.ts [hunk 5]
```diff
@@ -58,8 +58,32 @@ export abstract class HelpBase {
...
+    return this.config.topics.filter((topic: Config.Topic) => {
+      // it is assumed a topic has a child if it has children
+      const hasChild = this.config.topics.some(subTopic => subTopic.name.includes(`${topic.name}:`))
+      return hasChild
+    })
+  }
+
+  protected get filteredCommands() {
```
Comment (chadian) [inline src/index.ts:75]: change these from `filtered` to `sorted` prefixes

**Pull request merged**